File: leetcode/424.longest-repeating-character-replacement.py
# @leet start
from collections import deque
from string import ascii_uppercase

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        # NOTE: For this solution you really don't need a deque, just
        # keep track of the start index and end index using a "sliding
        # window" type technique. Also probably just move the
        # substr_is_valid logic outside of a function call? Lots of
        # little efficiencies here, but I got the time and memory
        # complexity I want, so I'm just going to move on instead of
        # making this super clean.

        # Keep track of the count of letters in the substring
        count_dict = {ch: 0 for ch in ascii_uppercase}

        # Define a function that tells us if the current substring is
        # valid using the count_dict. There's probably a better way to
        # do this but it's still O(1) so we're good.
        def substr_is_valid():
            max_count = max(count_dict.values())

            if len(substr_deque) > max_count + k:
                return False

            return True

        # Keep the substring in a deque
        substr_deque = deque()

        # Go through all valid substrings, find the best one
        best = 0

        for char in s:
            # Process new character
            substr_deque.append(char)
            count_dict[char] += 1

            # Pop out old characters until the substring is valid
            while not substr_is_valid():
                old_char = substr_deque.popleft()
                count_dict[old_char] -= 1

            # Update the result variable
            best = max(best, len(substr_deque))

        return best
# @leet end

File: leetcode/424.longest-repeating-character-replacement.js
// @leet start
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
  const CODE_POINT_A = "A".codePointAt(0);
  const counts = Array(26).fill(0);

  const isValid = () => {
    let maxCount = 0;
    let totalCount = 0;

    for (i = 0; i < 26; i++) {
      if (counts[i] > maxCount) {
        maxCount = counts[i];
      }

      totalCount += counts[i];
    }

    return totalCount - maxCount <= k;
  };

  let left = 0;
  let best = 0;

  for (let right = 0; right < s.length; right++) {
    // Add in right character
    counts[s[right].codePointAt(0) - CODE_POINT_A]++;

    // Shrink the window until valid
    while (!isValid()) {
      counts[s[left].codePointAt(0) - CODE_POINT_A]--;
      left++;
    }

    // Update best
    best = Math.max(best, right - left + 1);
  }

  return best;
};
// @leet end
